### PR TITLE
v1.12.0 geoips - VERSION RELEASE - plugin registry bug fixes, geoips install workflows

### PR DESCRIPTION
**Pull request by mindyls**
# Related Issues
required to close NRLMMD-GEOIPS/geoips#418
required to close NRLMMD-GEOIPS/geoips#402

# Reviewer Instructions
* Please confirm this PR is set up to merge from 'v1.12.0-release' branch into 'main' branch
* Review "Files changed" tab
* Confirm appropriate testing was completed for these updates (you may perform tests yourself, or review output included from another reviewer/assignee)

# Summary
Changes for version 1.12.0, uploaded with
1.12.0 update on repo geoips.
See release note updates in 'Files changed' tab

See NRLMMD-GEOIPS/geoips#418 for additional repositories included in this update.
See NRLMMD-GEOIPS/geoips#402 for dev updates included in this update.

# Testing Instructions
See NRLMMD-GEOIPS/geoips#418 for testing instructions.

# Output
See NRLMMD-GEOIPS/geoips#418 for test output.

# Post Merge Steps
Once this PR has been merged, v1.12.0
will be tagged/released on the geoips main branch

### CODE REVIEW

**Comment by jsolbrig** [comment]
I accidentally installed this in a conda environment with Python 3.9 and found that it no longer works for 3.9. Do we care about supporting 3.9?

If we don't want to support 3.9, we should update `geoips_utils.get_entry_points_group` and update pyproject.toml to require 3.10 or greater.

If we do want to support 3.9, we need to fix the following issue. That said, there may be additional problems.
```
Traceback (most recent call last):
  File "/home/jsolbrig/miniconda3/bin/run_procflow", line 8, in <module>
    sys.exit(main())
  File "/home/jsolbrig/miniconda3/lib/python3.9/site-packages/geoips/commandline/run_procflow.py", line 69, in main
    RETVAL = PROCFLOW(COMMAND_LINE_ARGS["filenames"], COMMAND_LINE_ARGS)
  File "/home/jsolbrig/miniconda3/lib/python3.9/site-packages/geoips/plugins/modules/procflows/single_source.py", line 1921, in call
    output_checker = output_checkers.get_plugin(output_product)
  File "/home/jsolbrig/miniconda3/lib/python3.9/site-packages/geoips/interfaces/module_based/output_checkers.py", line 1045, in get_plugin
    plug = super().get_plugin(name)
  File "/home/jsolbrig/miniconda3/lib/python3.9/site-packages/geoips/interfaces/base.py", line 718, in get_plugin
    module = find_entry_point(self.name, name)
  File "/home/jsolbrig/miniconda3/lib/python3.9/site-packages/geoips/geoips_utils.py", line 135, in find_entry_point
    for ep in get_entry_point_group(ep_namespace):
  File "/home/jsolbrig/miniconda3/lib/python3.9/site-packages/geoips/geoips_utils.py", line 39, in get_entry_point_group
    eps = metadata.entry_points()[group]
KeyError: 'geoips.output_checkers'
```

**Comment by jsolbrig** [comment]
In `interfaces.base`, `BaseModuleInterface.get_plugin()` has a section that says "# This is used! For output checkers at least." where we call `find_entry_point`. I thought we were done with needing to call `find_entry_point` now that we have the registry.

- Is this actually needed?
- If it is not needed, can we remove it?
- If it is needed, we should comment more thoroughly to explain why so there isn't confusion in the future.

**Comment by jsolbrig** [comment]
Probably create an issue for this. Just putting a note here since I'm in a rush.

It appears that `get_plugin` is the same for `BaseYamlInterface` and `BaseModuleInterface`. We should check whether this is true and, if so, combine them on `BaseInterface`.

**Review by lance-wilson — CHANGES_REQUESTED** [review]
Everything (that I have data for) passes for me, with the following exceptions:
* I'm getting "module 'geoips' has no attribute '__version__'" on most of the build_docs scripts.
* black still wants to reformat the "version.py" file for the non-geoips repos.
* There seem to be unrecognized pytest arguments (possibly my system is not set up correctly, though).

If the fix for the problem that Jeremy mentioned is supposed to be part of this release, then that also needs to get sorted out.

**Comment by AJThorpe** [comment]
Setup, installation, and testing went smoothly for me, ran on python 3.10.12:

base test:
```
Package: geoips_base
Total run time: 122 seconds
Number data types run: 3
Number data types failed: 0
Thu Feb  8 19:02:23 UTC 2024
```

Once other issues are resolved, I'll approve the merge.

**Comment by AJThorpe** [comment]
@jsolbrig
/cc @mindyls 
Will all packages fully port over to python 3.10?? I know clavrx was having issues with installation in python 3.10

**Comment by mindyls** [comment]
> @jsolbrig /cc @mindyls Will all packages fully port over to python 3.10?? I know clavrx was having issues with installation in python 3.10

I don't know if the clavrx issues are specifically with python 3.10, or just the libraries that get installed with conda when installing 3.10.

This all being said, I am ok with forcing 3.10 at this point.

File: geoips/geoips_utils.py
Comment (mindyls) [inline geoips/geoips_utils.py:39]: @jsolbrig This is where yours was erroring on python 3.9?  I think that must be a red herring, because that else is there specifically to support 3.9, there must be something else causing 3.9 to fail.

**Comment by mindyls** [comment]
> Everything (that I have data for) passes for me, with the following exceptions:
> 
> * I'm getting "module 'geoips' has no attribute '__version__'" on most of the build_docs scripts.
> * black still wants to reformat the "version.py" file for the non-geoips repos (possibly these are leftover files that I need to delete, though).
> * There seem to be unrecognized pytest arguments (possibly my system is not set up correctly, though).
> 
> If the fix for the problem that Jeremy mentioned is supposed to be part of this release, then that also needs to get sorted out.

I would say ignore those issues - actions workflows are passing, must be something in your environment.

File: geoips/geoips_utils.py
Comment (mindyls) [reply geoips/geoips_utils.py:39]: https://github.com/NRLMMD-GEOIPS/geoips/pull/419#issuecomment-1924466402

File: geoips/interfaces/base.py
Comment (mindyls) [inline geoips/interfaces/base.py:717]: https://github.com/NRLMMD-GEOIPS/geoips/pull/419#issuecomment-1924471870

Can we ignore this for now?  Looks like we need to open 2 issues, then this is good to go?  Look into if we can remove this (I'm guessing no, since I *just* added it...), and look into whether get_plugin can be combined (probably, since we have officially moved to a single source/product for for product plugins). @jsolbrig

Comment (mindyls) [reply geoips/interfaces/base.py:717]: https://github.com/NRLMMD-GEOIPS/geoips/issues/435

Comment (lance-wilson) [reply geoips/interfaces/base.py:717]: @mindyls Was there some reason this was different in the open-source GeoIPS?  I'm sure we got rid of the output_checkers special case internally.

**Review by lance-wilson — APPROVED** [review]

File: geoips/geoips_utils.py
Comment (mindyls) [reply geoips/geoips_utils.py:39]: https://github.com/NRLMMD-GEOIPS/geoips/pull/419#issuecomment-1934885067
I think we are good to go with removing support for 3.9 at this point. We will deal with additional errors as they arise.

File: geoips/interfaces/base.py
Comment (mindyls) [reply geoips/interfaces/base.py:717]: I expect the change was just made after 1.12.0 was bundled up for release.  I say resolve this, and it will be gone in the next release.

File: geoips/geoips_utils.py
Comment (mindyls) [reply geoips/geoips_utils.py:39]: https://github.com/NRLMMD-GEOIPS/geoips/pull/419#issuecomment-1934875033
Note there may be issues with CLAVR-x build, etc, but we will deal with those separately (that is an external package, not actually part of the geoips plugin ecosystem)

File: geoips/interfaces/base.py
Comment (mindyls) [reply geoips/interfaces/base.py:717]: https://github.com/NRLMMD-GEOIPS/geoips/pull/419#issuecomment-1927934980
There is lance's original comment RE: getting rid of that special case, for reference.

Comment (mindyls) [reply geoips/interfaces/base.py:717]: https://github.com/NRLMMD-GEOIPS/geoips/pull/419#issuecomment-1924471870 <-- Jeremy's original comment, for reference.  Hidden as duplicate now.

File: geoips/interfaces/base.py
Comment (mindyls) [inline geoips/interfaces/base.py:714]: https://github.com/NRLMMD-GEOIPS/geoips/issues/436
https://github.com/NRLMMD-GEOIPS/geoips/pull/419#issuecomment-1924473740

It appears that get_plugin is the same for BaseYamlInterface and BaseModuleInterface. We should check whether this is true and, if so, combine them on BaseInterface.  

get_plugin is defined right here on base.py, so not sure why they are subclassed on YAML and Module if they are the same..

Comment (mindyls) [reply geoips/interfaces/base.py:714]: Since I opened #436 to investigate this, I think we should resolve this conversation.

File: .config/flake8
Comment (jsolbrig) [inline .config/flake8:12]: Would it be better for this to be directly in the file itself rather than at the top level here?

File: geoips/dev/product.py
Comment (jsolbrig) [inline geoips/dev/product.py:1]: Wait, is this still used?! Have we not finished the new interfaces?

File: geoips/geoips_utils.py
Comment (jsolbrig) [reply geoips/geoips_utils.py:39]: Yes, this is where my code was failing on 3.9. Should we create an issue to remove any references to 3.9? There are references elsewhere, too, especially in the CI.

File: geoips/dev/product.py
Comment (jsolbrig) [reply geoips/dev/product.py:1]: @mindy answered this. Yes, it is still used by the config-based procflow. It might be worth considering moving it to somewhere closer to the config-based procflow, but should leave it for now.

File: .config/flake8
Comment (mindyls) [reply .config/flake8:12]: I think perhaps in both.  I'll change it on my internal updates now.

File: geoips/plugins/modules/readers/seviri_hrit.py
Comment (jsolbrig) [inline geoips/plugins/modules/readers/seviri_hrit.py:39]: Why do the changes on lines 31 and 37 appear to be opposite changes? Is there a preference here?

File: setup/config_geoips
Comment (jsolbrig) [inline setup/config_geoips:12]: Just for my information, what does this do? I read the documentation on creatordate, but I still don't understand how this change effects our geoips versioning scheme.

File: geoips/geoips_utils.py
Comment (jsolbrig) [reply geoips/geoips_utils.py:39]: Created an issue relating to this. Resolving for now.

File: setup/config_geoips
Comment (mindyls) [reply setup/config_geoips:12]: It is just pulling the geoips version from the tags - it does not have any impact on actually setting the versions, etc.

File: geoips/plugins/modules/readers/seviri_hrit.py
Comment (mindyls) [reply geoips/plugins/modules/readers/seviri_hrit.py:39]: I suppose relative imports is probably preferable to imports with the full path?  Though I'm not sure it matters too much?  I think at one stage of my bug fixing with the plugin registries, etc one or the other didn't work, but it appears now that both work if there is one of each in here...  I think ignore this for now, and resolve the conversation.

File: geoips/plugins/modules/filename_formatters/utils/tc_file_naming.py
Comment (mindyls) [reply geoips/plugins/modules/filename_formatters/utils/tc_file_naming.py:25]: I can comment these on my internal update I am doing right now.

File: geoips/interfaces/base.py
Comment (jsolbrig) [reply geoips/interfaces/base.py:717]: Okay, sounds good. I created #440 to consider combining get_plugin on `BaseInterface`.

File: geoips/plugins/modules/filename_formatters/utils/tc_file_naming.py
Comment (jsolbrig) [reply geoips/plugins/modules/filename_formatters/utils/tc_file_naming.py:25]: This will come in 1.12.1.

File: geoips/plugins/modules/readers/seviri_hrit.py
Comment (jsolbrig) [reply geoips/plugins/modules/readers/seviri_hrit.py:39]: Okay, I'll resolve since it works. Just odd to make the opposite change on lines in the same file.

File: setup/config_geoips
Comment (jsolbrig) [reply setup/config_geoips:12]: Okay, thanks!

File: geoips/interfaces/base.py
Comment (jsolbrig) [reply geoips/interfaces/base.py:714]: Ah, oops, I opened an issue for this too. I'll close it as duplicate and close this.

**Review by jsolbrig — APPROVED** [review]
All conversations are resolved. All tests pass.

**Pull request merged**
